### PR TITLE
Fix SendInput handling

### LINKED ISSUE #7438: The keyboard shortcut: "ctrl + j = ←", "ctrl + i = ↑", "ctrl + l = →", "ctrl + k = ↓" is broken.
# Environment

```none
Windows build number: 10.0.20201.0 Microsoft Windows NT 10.0.20201.0
Windows Terminal version: 1.2.2381.0

Any other software?
```

# Steps to reproduce

I have set up the replacement of keyboard shortcuts in "PowerToys" using the "Keyboard Manager" option: 

```
ctrl+j = ← 
ctrl+i = ↑ 
ctrl+l = → 
ctrl+k = ↓
```

This is very convenient — it will allow you to hold both hands in the center of the keyboard without having to often use the arrows on the side of the working area on the keyboard. 

In v1.2.2381.0 version these key combinations are broken. 

But the combination: "ctrl + h = Backspace" works.

I had to roll back to v1.2.2234.0 where the keyboard shortcuts work. 

There is also a problem with keyboard shortcuts in Windows Terminal Preview v1.3.2382.0.

# What I expect

By pressing the key combination ctrl+j I expect to get the command "←". 

By pressing the ctrl+i key combination I expect to get the "↑" command. 

By pressing the ctrl+l key combination I expect to get the command "→". 

By pressing the ctrl+k key combination, I expect to receive the command "↓". 

# What happening actually

By pressing "ctrl+j", "ctrl+i", "ctrl+l", "ctrl+k", Windows Terminal does not react in any way — nothing happens. In other programs, the specified keyboard shortcuts work correctly. In the Windows Terminal v1.2.2234.0 version, the specified keyboard shortcuts also worked correctly.

### LINKED ISSUE #7495: Arrow keys not working on Touchscreen (on-screen) keyboard
# Environment

```none
Windows build number: 10.0.20201.0
Windows Terminal version (if applicable): 1.3.2382.0
```

# Steps to reproduce
1. Open Windows Terminal Pre release on a computer with touch screen.
2. Popup the "on-screen" keyboard
3. Set the keyboard to "full size", to get arrow keys
4. Enter a command, try to click arrow left/right
5. Run the command and click up

This is tested both in PowerShell and WSL 2 through Windows Terminal

# Expected behavior

I would expect the arrow keys to work, for example for going back in command history, or to use arrow keys in Vim or similar.

# Actual behavior

The arrow keys is not working, nothing happens when you click them. You can try to open e.g. Windows PowerShell directly, where it will work.

### LINKED ISSUE #7843: WT programs do not respond to arrow keys sent from the SendKeys API
# Environment

```none
Windows build number: 10.0.18362.0
Windows Terminal version 1.3.2651.0:

Any other software?

A C# program using the SendKeys API
```

# Steps to reproduce

Write a program in C# using the SendKeys API to emulate keypresses. Use this program to attempt to send arrow keys to a program running in Windows Terminal.

# Expected behavior

The program in Windows Terminal should respond as if an arrow key has been pressed. In Powershell or CMD or most shells, the cursor should move according to the arrow key that is pressed.

# Actual behavior

The cursor blinks as if a key has been pressed, but will not move. The program in WT does not respond in any other way. The up and down arrow keys sent by SendKeys WILL work in the dropdown menu however. It seems like programs running in WT will respond to most other keys appropriately.

### PR DESCRIPTION
**Pull request by lhecker**
While not explicitly permitted, a wide range of software (including
Windows' own touch keyboard) sets the `wScan` member of the `KEYBDINPUT`
structure to 0, resulting in `scanCode` being 0 as well.  In these
situations we'll now use the `vkey` to get a `scanCode`.

Validation
----------
* AutoHotkey
  * Use a keyboard layout with `AltGr` key
  * Execute the following script:
    ```ahk
    #NoEnv
    #Warn
    SendMode Input
    SetWorkingDir %A_ScriptDir%
    <^>!8::SendInput {Raw}»
    ```
  * Press `AltGr+8` while the Terminal is in the foreground
  * Ensure » is being echoed ✔️
* PowerToys
  * Add a `Ctrl+I -> ↑ (up arrow)` keyboard shortcut
  * Press `Ctrl+I` while the Terminal is in the foreground
  * Ensure the shell history is being navigated backwards ✔️
* Windows Touch Keyboard
  * Right-click or tap and hold the taskbar and select "Show touch
    keyboard" button
  * Open touch keyboard
  * Ensure keyboard works like a regular keyboard ✔️
  * Ensure unicode characters are echoed on the Terminal as well (except
    for Emojis) ✔️

Closes #7438
Closes #7495
Closes #7843

### CODE REVIEW

File: src/cascadia/UnitTests_TerminalCore/InputTest.cpp [hunk 1]
```diff
@@ -71,9 +71,7 @@ namespace TerminalCoreUnitTests
     {
         // Certain applications like AutoHotKey and its keyboard remapping feature,
         // send us key events using SendInput() whose values are outside of the valid range.
-        // We don't want to handle those events as we're probably going to get a proper followup character event.
         VERIFY_IS_FALSE(term.SendKeyEvent(0, 123, {}, true));
         VERIFY_IS_FALSE(term.SendKeyEvent(255, 123, {}, true));
-        VERIFY_IS_FALSE(term.SendKeyEvent(123, 0, {}, true));
```
Comment (lhecker) [inline src/cascadia/UnitTests_TerminalCore/InputTest.cpp:77]: The already minimal test coverage for this particular method is being reduced even further. 😕
Unfortunately I don't believe there's any valid `vkey` on "regular" computers that maps to a zero scanCode using `MapVirtualKeyW`.

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 2]
```diff
@@ -471,15 +471,20 @@ bool Terminal::SendKeyEvent(const WORD vkey,
...
+    {
+        return false;
+    }
+
+    // While not explicitly permitted, a wide range of software, including Windows' own touch keyboard,
+    // sets the wScan member of the KEYBDINPUT structure to 0, resulting in scanCode being 0 as well.
+    // --> Alternatively get the scanCode from the vkey if possible.
+    // GH#7495
+    const auto sc = scanCode ? scanCode : _ScanCodeFromVirtualKey(vkey);
+    if (sc == 0)
```
Comment (lhecker) [inline src/cascadia/TerminalCore/Terminal.cpp:487]: I've been thinking for a while whether I should allow `sc` to be 0 or not. I've decided for the latter, because:
WT is already broken and if we un-break WT by just enough for this to be fixed, we can keep more defensive mechanisms, allowing us to rely more on certain properties, like... `scanCode` never being zero in `TerminalInput`.

I think it should be fine to keep this if-condition. 🙂 (...until some weird Lumia 950 XL with full fledged Windows 10 Pro proves us wrong 😂)

**Comment by lhecker** [comment]
If anyone has any suggestions which corner cases I should additionally test, _please_ let me know.
I don't like breaking stuff and I'd really like to "unbreak" all the corner cases now. 😅

File: src/cascadia/TerminalCore/Terminal.cpp (hunk 2 above)
Comment (lhecker) [reply src/cascadia/TerminalCore/Terminal.cpp:487]: I think it's important I mention the reason we need `scanCode`s here again (because this topic is kinda complicated)...
When we receive a key event we must decide whether we consume the event or not. If we do consume it we'll not receive a follow-up character event no matter what. We'd like to only consume non-character key events, because the character events contain proper unicode characters and the key events don't.
In order to decide this we use `ToUnicodeEx` which translates a `vkey` and some modifier keys to one or more characters. If it spits out characters we ignore the key and wait for an appropriate character event.
Unfortunately there can be multiple meanings to a single `vkey` depending on the keyboard state and depending on the key in question that was pressed. Two or more keys might map to the same `vkey` but have different `scanCode`s.
Due to this `ToUnicodeEx` asks for a `scanCode` allowing it to correctly map unicode characters.

Comment (DHowett) [reply src/cascadia/TerminalCore/Terminal.cpp:487]: This is an excellent description. Thanks!

**Review by DHowett — APPROVED** [review]
This looks great and very well-reasoned to me,

**Review by zadjii-msft — APPROVED** [review]
Alright, so after enough coffee, this doesn't actually look all that scary. I quite like the clever fallback of looking up the scan code from the vkey, since that's likely been set correctly. 

Sorry for the delay in reviewing!

**Comment by zadjii-msft** [comment]
I'm gonna check to see if I can repro #8045, and see if this fixes that. Turns out my dev box is still on, despite no one being in the office for the last 8 months 😆

**Comment by zadjii-msft** [comment]
I couldn't repro #8045, so I'm just gonna merge this, and hope we figure out the root cause of that issue separately.

**Comment by lhecker** [comment]
@D3nbot Which service are you talking about? None of the 3 issues this PR closed mention anything related to "service". From what I can tell, this PR also still seems to work for me.
